File: android/src/main/java/com/scandit/datacapture/flutter/core/ScanditFlutterDataCaptureCorePlugin.java
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2024- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.core;

import androidx.annotation.NonNull;

import com.scandit.datacapture.flutter.core.extensions.MethodChannelExtensions;
import com.scandit.datacapture.flutter.core.ui.ScanditPlatformViewFactory;
import com.scandit.datacapture.flutter.core.utils.FlutterEmitter;
import com.scandit.datacapture.frameworks.core.CoreModule;
import com.scandit.datacapture.frameworks.core.FrameworkModule;
import com.scandit.datacapture.frameworks.core.listeners.FrameworksDataCaptureContextListener;
import com.scandit.datacapture.frameworks.core.listeners.FrameworksDataCaptureViewListener;
import com.scandit.datacapture.frameworks.core.listeners.FrameworksFrameSourceListener;
import com.scandit.datacapture.frameworks.core.locator.DefaultServiceLocator;
import com.scandit.datacapture.frameworks.core.locator.ServiceLocator;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodChannel;

import java.lang.ref.WeakReference;
import java.util.concurrent.locks.ReentrantLock;

public class ScanditFlutterDataCaptureCorePlugin implements FlutterPlugin, ActivityAware {

    private static final ReentrantLock lock = new ReentrantLock();

    private final static FlutterEmitter coreEmitter = new FlutterEmitter(DataCaptureCoreMethodHandler.EVENT_CHANNEL_NAME);

    private final ServiceLocator<FrameworkModule> serviceLocator = DefaultServiceLocator.getInstance();

    private MethodChannel methodChannel;
    private WeakReference<FlutterPlugin.FlutterPluginBinding> flutterPluginBinding = new WeakReference<>(null);

    @Override
    public void onAttachedToEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        flutterPluginBinding = new WeakReference<>(binding);
        setupModules(binding);
        setupMethodChannels(binding);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        flutterPluginBinding = new WeakReference<>(null);
        disposeModules();
        disposeMethodChannels();
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        setupEventChannels();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        // NOOP
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        // NOOP
    }

    @Override
    public void onDetachedFromActivity() {
        disposeEventChannels();
    }

    private void setupEventChannels() {
        FlutterPluginBinding binding = flutterPluginBinding.get();
        if (binding != null) {
            coreEmitter.addChannel(binding.getBinaryMessenger());
        }
    }

    private void disposeEventChannels() {
        FlutterPluginBinding binding = flutterPluginBinding.get();
        if (binding != null) {
            coreEmitter.removeChannel(binding.getBinaryMessenger());
        }
    }

    private void setupModules(FlutterPlugin.FlutterPluginBinding binding) {
        lock.lock();
        try {
            CoreModule coreModule = (CoreModule) serviceLocator.resolve(CoreModule.class.getName());
            if (coreModule != null) return;

            coreModule = CoreModule.create(
                    new FrameworksFrameSourceListener(coreEmitter),
                    new FrameworksDataCaptureContextListener(coreEmitter),
                    new FrameworksDataCaptureViewListener(coreEmitter)
            );
            coreModule.onCreate(binding.getApplicationContext());
            coreModule.registerDataCaptureContextListener();
            coreModule.registerDataCaptureViewListener();
            coreModule.registerFrameSourceListener();

            serviceLocator.register(coreModule);
        } finally {
            lock.unlock();
        }
    }

    private void disposeModules() {
        lock.lock();
        try {
            FrameworkModule module = serviceLocator.remove(CoreModule.class.getName());
            if (module != null) {
                module.onDestroy();
            }
        } finally {
            lock.unlock();
        }
    }

    private void setupMethodChannels(@NonNull FlutterPluginBinding binding) {
        DataCaptureCoreMethodHandler dataCaptureCoreMethodHandler = new DataCaptureCoreMethodHandler(serviceLocator);
        binding.getPlatformViewRegistry().registerViewFactory(
                "com.scandit.DataCaptureView",
                new ScanditPlatformViewFactory(serviceLocator)
        );
        methodChannel = MethodChannelExtensions.getMethodChannel(binding, DataCaptureCoreMethodHandler.METHOD_CHANNEL_NAME);
        methodChannel.setMethodCallHandler(dataCaptureCoreMethodHandler);
    }

    private void disposeMethodChannels() {
        if (methodChannel != null) {
            methodChannel.setMethodCallHandler(null);
            methodChannel = null;
        }
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/flutter/core/ScanditFlutterDataCaptureCorePlugin.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2020- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.core

import com.scandit.datacapture.flutter.core.extensions.getMethodChannel
import com.scandit.datacapture.flutter.core.utils.FlutterEmitter
import com.scandit.datacapture.frameworks.core.CoreModule
import com.scandit.datacapture.frameworks.core.listeners.FrameworksDataCaptureContextListener
import com.scandit.datacapture.frameworks.core.listeners.FrameworksDataCaptureViewListener
import com.scandit.datacapture.frameworks.core.listeners.FrameworksFrameSourceListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import java.lang.ref.WeakReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/** ScanditFlutterDataCaptureCorePlugin. */
class ScanditFlutterDataCaptureCorePlugin :
    FlutterPlugin, ActivityAware {

    companion object {
        @JvmStatic
        private val lock = ReentrantLock()

        @JvmStatic
        private var isPluginAttached = false

        private const val EVENT_CHANNEL_NAME =
            "com.scandit.datacapture.core/event_channel"
        private const val METHOD_CHANNEL_NAME =
            "com.scandit.datacapture.core/method_channel"
        private const val DATACAPTURE_VIEW_ID = "com.scandit.DataCaptureView"
    }

    private var methodChannel: MethodChannel? = null

    private var coreModule: CoreModule? = null

    private var flutterPluginBinding: WeakReference<FlutterPluginBinding?> = WeakReference(null)

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        flutterPluginBinding = WeakReference(binding)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        flutterPluginBinding = WeakReference(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        onAttached()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetached()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttached()
    }

    override fun onDetachedFromActivity() {
        onDetached()
    }

    private fun onAttached() {
        lock.withLock {
            if (isPluginAttached) return

            val flutterBinding = flutterPluginBinding.get() ?: return

            setupModule(flutterBinding)

            isPluginAttached = true
        }
    }

    private fun onDetached() {
        lock.withLock {
            disposeModules()
            isPluginAttached = false
        }
    }

    private fun setupModule(binding: FlutterPluginBinding) {
        val eventEmitter = FlutterEmitter(
            EventChannel(
                binding.binaryMessenger,
                EVENT_CHANNEL_NAME
            )
        )

        val coreModule = CoreModule(
            FrameworksFrameSourceListener(
                eventEmitter
            ),
            FrameworksDataCaptureContextListener(
                eventEmitter
            ),
            FrameworksDataCaptureViewListener(
                eventEmitter
            )
        ).also {
            it.onCreate(binding.applicationContext)
            it.registerDataCaptureContextListener()
            it.registerDataCaptureViewListener()
            it.registerFrameSourceListener()
        }

        val dataCaptureCoreMethodHandler = DataCaptureCoreMethodHandler(
            coreModule
        )

        this.coreModule = coreModule

        binding.platformViewRegistry.registerViewFactory(
            DATACAPTURE_VIEW_ID,
            ScanditPlatformViewFactory(coreModule)
        )
        methodChannel = binding.getMethodChannel(METHOD_CHANNEL_NAME).also {
            it.setMethodCallHandler(dataCaptureCoreMethodHandler)
        }
    }

    private fun disposeModules() {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        coreModule?.onDestroy()
        coreModule = null
    }
}
